app/app.go

```diff
@@ -478,17 +484,36 @@ func NewStrideApp(
 	)
 	ratelimitModule := ratelimitmodule.NewAppModule(appCodec, app.RatelimitKeeper)
 
+	// Initialize the packet forward middleware Keeper
+	// It's important to note that the PFM Keeper must be initialized before the Transfer Keeper
+	app.PacketForwardKeeper = packetforwardkeeper.NewKeeper(
+		appCodec,
+		keys[packetforwardtypes.StoreKey],
+		// app.GetSubspace(packetforwardtypes.ModuleName), TODO: why is this in the integration docs but not in code?
+		app.TransferKeeper, // will be zero-value here, reference is set later on with SetTransferKeeper.
+		app.IBCKeeper.ChannelKeeper,
+		app.DistrKeeper,
+		app.BankKeeper,
+		app.IBCKeeper.ChannelKeeper,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(), // TODO: this is not in the integration docs, but a comment says the authority should be the gov module
+	)
+
 	// Create Transfer Keepers
 	app.TransferKeeper = ibctransferkeeper.NewKeeper(
 		appCodec, keys[ibctransfertypes.StoreKey],
 		app.GetSubspace(ibctransfertypes.ModuleName),
+		// TODO: Do we need to import the app.PacketForwardKeeper here?
 		app.RatelimitKeeper, // ICS4Wrapper
```

Comment:
based on the below stack ordering, I think we'll want to set app.PacketForwardKeeper here and then set app.RatelimitKeeper as the ICS4Wrapper of the PacketForwardKeeper
```go
app.RatelimitKeeper = (
   app.ChannelKeeper,
)   

app.PacketForwardKeeper = (
    app.RatelimitKeeper,
)

app.TransferKeeper = (
   app.PacketForwardKeeper
)
```

Reply:
I see there's a todo above this line "Implement ICS4Wrapper in Records and pass records keeper here" - is that relevant?

I'm not too familiar with the `ICS4Wrapper`, are you? I implemented based on your suggestion.

Reply:
The wrapper is how you define traffic _up_ the stack. 

Traffic _down_ the stack is defined [[here](https://github.com/Stride-Labs/stride/blob/main/app/app.go#L703-L713)] based on the ordering with which each middleware is wrapped. But traffic _up_ the stack travels through these ICS4Wrappers. 

Right now, our upwards stack is technically just `Transfer` [[→](https://github.com/Stride-Labs/stride/blob/main/app/app.go#L485)] `RateLimit` [[→](https://github.com/Stride-Labs/stride/blob/main/app/app.go#L477)] `Core IBC (Channel)` (no records or autopilot)

---

As to the TODO you're referencing, that was originally added because it technically felt more more correct to have the up and down stack be mirror images of each other. But on second though, considering records and Autopilot doesn't need to override anything during the ICS4 calls (SendPacket, WriteAcknowledgement), I don't think it's necessary to include them in the stack. 

Curious to hear your thoughts though!

---

All that said, at the very least, we need to make sure to include PFM in this upward stack (per original comment)

Reply:
Ah, gotcha. Thanks for explaining that. 

I agree, they should mirror each other but don't need to since Autopilot doesn't need to override anything during the ICS4 calls.

> All that said, at the very least, we need to make sure to include PFM in this upward stack (per original comment)

Is [this line](https://github.com/Stride-Labs/stride/pull/889/files#diff-0f1d2976054440336a576d47a44a37b80cdf6701dd9113012bce0e3c425819b7R503) sufficient?

Reply:
Yup! that and [this](https://github.com/Stride-Labs/stride/pull/889/files#diff-0f1d2976054440336a576d47a44a37b80cdf6701dd9113012bce0e3c425819b7R495) are all that's needed